modules/index/src/test/java/org/apache/ignite/internal/index/IndexBuildControllerTest.java

```diff
@@ -0,0 +1,280 @@
+    }
+
+    private void setPrimaryReplicaForOneSecond(int partitionId, String leaseholder, HybridTimestamp startTime) {
+        CompletableFuture<ReplicaMeta> replicaMetaFuture = completedFuture(replicaMetaForOneSecond(leaseholder, startTime));
+
+        assertThat(placementDriver.setPrimaryReplicaMeta(0, replicaId(partitionId), replicaMetaFuture), willCompleteSuccessfully());
+    }
+
+    private int tableId() {
+        return getTableIdStrict(catalogManager, TABLE_NAME, clock.nowLong());
+    }
+
+    private int indexId(String indexName) {
+        return getIndexIdStrict(catalogManager, indexName, clock.nowLong());
+    }
+
+    private TablePartitionId replicaId(int partitionId) {
+        return new TablePartitionId(tableId(), partitionId);
+    }
+
+    private static TestReplicaMeta replicaMetaForOneSecond(String leaseholder, HybridTimestamp startTime) {
+        HybridTimestamp expirationTime = new HybridTimestamp(startTime.getPhysical() + 1_000, startTime.getLogical());
+
+        return new TestReplicaMeta(leaseholder, startTime, expirationTime);
+    }
+
+    private IndexBuildController newIndexBuildController() {
+        return new IndexBuildController(indexBuilder, indexManager, catalogManager, clusterService, placementDriver, clock);
+    }
+
+    private static class TestPlacementDriver extends AbstractEventProducer<PrimaryReplicaEvent, PrimaryReplicaEventParameters> implements
+            PlacementDriver {
+        private final Map<ReplicationGroupId, CompletableFuture<ReplicaMeta>> primaryReplicaMetaFutureById = new ConcurrentHashMap<>();
+
+        @Override
+        public CompletableFuture<ReplicaMeta> awaitPrimaryReplica(
+                ReplicationGroupId groupId,
+                HybridTimestamp timestamp,
+                long timeout,
+                TimeUnit unit
+        ) {
+            return primaryReplicaMetaFutureById.get(groupId);
+        }
+
+        @Override
+        public CompletableFuture<ReplicaMeta> getPrimaryReplica(ReplicationGroupId replicationGroupId, HybridTimestamp timestamp) {
+            return primaryReplicaMetaFutureById.get(replicationGroupId);
+        }
+
+        @Override
+        public CompletableFuture<Void> previousPrimaryExpired(ReplicationGroupId grpId) {
+            throw new UnsupportedOperationException();
+        }
+
+        CompletableFuture<Void> setPrimaryReplicaMeta(
+                long causalityToken,
+                TablePartitionId replicaId,
+                CompletableFuture<ReplicaMeta> replicaMetaFuture
+        ) {
+            primaryReplicaMetaFutureById.put(replicaId, replicaMetaFuture);
+
+            return replicaMetaFuture.thenCompose(replicaMeta -> fireEvent(
+                    PrimaryReplicaEvent.PRIMARY_REPLICA_ELECTED,
+                    new PrimaryReplicaEventParameters(causalityToken, replicaId, replicaMeta.getLeaseholder())
+            ));
+        }
+    }
+
+    private static class TestReplicaMeta implements ReplicaMeta {
```

Comment:
Would it be simpler to use `Lease` class instead of the test implementation?

Reply:
Fix it